source/AI.cpp

```diff
@@ -3385,35 +3385,128 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
+		// It will be focusing on hostiles before allies. Holding `Shift`
+		// results in boarding only player-owned escorts in need of assistance.
 		shared_ptr<const Ship> target = ship.GetTargetShip();
-		if(!target || !CanBoard(ship, *target) || (shift && !target->IsYours()))
+		if(!target || (target && activeCommands.Has(Command::WAIT)) || !CanBoard(ship, *target)
+				|| (shift && !target->IsYours()))
 		{
 			if(shift)
 				ship.SetTargetShip(shared_ptr<Ship>());
 
-			double closest = numeric_limits<double>::infinity();
 			bool foundEnemy = false;
-			bool foundAnything = false;
-			for(const shared_ptr<Ship> &other : ships)
-				if(CanBoard(ship, *other))
+			const auto boardingPriority = Preferences::GetBoardingPriority();
+
+			auto strategy = [&]() noexcept -> function<double(Ship &)>
+			{
+				Point current = ship.Position();
+				switch(boardingPriority)
 				{
-					if(shift && !other->IsYours())
-						continue;
+					case Preferences::BoardingPriority::cost:
+						return [this, &ship](Ship &other) noexcept -> double
+						{
+							// Use the exact cost if the ship was scanned, otherwise use an estimation.
+							return this->Has(ship, other.shared_from_this(), ShipEvent::SCAN_OUTFITS) ?
+								other.Cost() : (other.ChassisCost() * 2.);
+						};
+					case Preferences::BoardingPriority::mixed:
+						return [this, &ship, current](Ship &other) noexcept -> double
+						{
+							double cost = this->Has(ship, other.shared_from_this(), ShipEvent::SCAN_OUTFITS) ?
+								other.Cost() : (other.ChassisCost() * 2.);
+							// Even if we divide by 0, doubles can contain and handle infinity,
+							// and we should definitely board that one then.
+							return cost * cost / (current.DistanceSquared(other.Position()) + 0.1);
+						};
+					// Default to distance priorities (the default setting).
+					default:
+					case Preferences::BoardingPriority::distance:
+						return [current](Ship &other) noexcept -> double
+						{
+							return current.DistanceSquared(other.Position());
+						};
+				}
+			}();
+
+			using shipValue = pair<Ship *, double>;
+			auto boardable = vector<shipValue>{};
+
+			auto fillBoardable = [&ship, &foundEnemy, &boardable, &strategy](Ship &other) noexcept
+			{
+				if(CanBoard(ship, other))
+					boardable.emplace_back(&other, strategy(other));
+			};
+
+			if(shift)
+			{
+				for(Ship *escort : governmentRosters[ship.GetGovernment()])
+					if(escort->GetSystem() == ship.GetSystem())
+						fillBoardable(*escort);
+			}
+			else
+			{
+				auto ships = GetShipsList(ship, true);
+				boardable.reserve(ships.size());
+				// First check if we can board enemy ships, then allies.
+				for(Ship *enemy : ships)
+					fillBoardable(*enemy);
+				// The current target is not considered by GetShipsList.
+				if(target)
+					fillBoardable(*const_cast<Ship *>(target.get()));
+
+				if(boardable.empty())
+				{
+					ships = GetShipsList(ship, false);
+					if(ships.size() > boardable.size())
```

Comment:
If we want to `reserve()`, then we should check against `boardable.capacity()`, or even better, just call `reserve()` directly without checking `boardable.capacity()`, since reserve has the capacity check already built-in.

Reply:
I've never used reserve before so yeah idk I guess I'll do this

Reply:
I think everything is good now